### PR TITLE
[REF] *:replace moment usages with luxon

### PR DESCRIPTION
**Pull request by pipu-odoo**
luxon and moment are both used in the solution, but these two libraries facilitate the manipulation of dates. It was decided to replace all uses of moment with luxon so we can then remove moment.js from the code and lighten the assets.

task-3391739

PR Enterprise : https://github.com/odoo/enterprise/pull/44051

### CODE REVIEW

File: addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js [hunk 1]
```diff
@@ -365,7 +367,8 @@ export class TicketScreen extends Component {
         return this._getOrderList().filter(predicate);
     }
     getDate(order) {
-        return deserializeDate(order.validation_date).toFormat("yyyy-MM-dd HH:mm a");
+        order.validation_date = order.validation_date || DateTime.now();
+        return DateTime.fromISO(order.validation_date).toFormat("yyyy-MM-dd HH:mm a");
```
Comment (pipu-odoo) [inline addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js:371]: @loukwets This is the fix as we discussed.

File: addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js [hunk 2]
```diff
@@ -365,7 +367,15 @@ export class TicketScreen extends Component {
...
-        return deserializeDate(order.validation_date).toFormat("yyyy-MM-dd HH:mm a");
+        let { validation_date } = order;
+        if (typeof validation_date === "string") {
+            validation_date = new Date(validation_date);
+        }
+        validation_date =
+            validation_date instanceof Date && !isNaN(validation_date.valueOf())
+                ? validation_date
+                : DateTime.local().toJSDate();
+        return DateTime.fromJSDate(order.validation_date).toFormat("yyyy-MM-dd HH:mm a");
```
Comment (pipu-odoo) [inline addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js:378]: @loukwets The fix as we discussed

File: addons/pos_restaurant/static/src/overrides/models/models.js [hunk 3]
```diff
@@ -29,7 +30,9 @@ patch(Order.prototype, {
         super.init_from_JSON(...arguments);
         if (this.pos.config.module_pos_restaurant) {
             this.tableId = json.table_id;
-            this.validation_date = moment.utc(json.creation_date).local().toDate();
+            const format = "yyyy-MM-dd HH:mm:ssZZ";
+            json.creation_date = json.creation_date || DateTime.now().toFormat(format); //ex : json.creation_date = "2023-08-25 13:24:16+02:00"
+            this.validation_date = DateTime.fromFormat(json.creation_date, format).toJSDate();
```
Comment (brboi) [inline addons/pos_restaurant/static/src/overrides/models/models.js:35]: I do not understand why you need to replace `json.creation_date` here
We will discuss this IRL tomorrow.

File: addons/website_links/static/src/js/website_links_charts.js [hunk 4]
```diff
@@ -28,11 +32,10 @@ var BarChart = publicWidget.Widget.extend({
...
         // Fill data for each day (with 0 click for days without data)
         var clicksArray = [];
-        var beginDateCopy = this.beginDate;
-        for (var i = 0; i < this.number_of_days; i++) {
-            var dateKey = beginDateCopy.format('YYYY-MM-DD');
+        for (var i = 0; i <= this.number_of_days; i++) {
+            var dateKey = this.beginDate.toFormat("yyyy-MM-dd");
             clicksArray.push([dateKey, (dateKey in this.dates) ? this.dates[dateKey] : 0]);
-            beginDateCopy.add(1, 'days');
+            this.beginDate = this.beginDate.plus({ days: 1 });
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website_links/static/src/js/website_links_charts.js:38]: Yes, I've checked without and not works.

File: addons/website_links/static/src/js/website_links_charts.js [hunk 5]
```diff
@@ -166,29 +169,29 @@ publicWidget.registry.websiteLinksCharts = publicWidget.Widget.extend({
                 // This is a trick to get the date without the local formatting.
                 // We can't simply do .locale("en") because some Odoo languages
                 // are not supported by moment.js (eg: Arabic Syria).
-                const date = moment(
+                const date = DateTime.fromFormat(
                     _clicksByDay[i]["__domain"].find((el) => el.length && el.includes(">="))[2]
-                        .split(" ")[0], "YYYY MM DD"
+                        .split(" ")[0], "yyyy-MM-dd"
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website_links/static/src/js/website_links_charts.js:174]: Due to 
moment("22-01-2022", "DD MM YYYY").isValid() 
> true 
luxon.DateTime.fromFormat("22-01-2022", "dd MM yyyy").isValid
> false

File: addons/pos_restaurant/static/src/overrides/models/models.js (hunk 3 above)
Comment (pipu-odoo) [reply addons/pos_restaurant/static/src/overrides/models/models.js:35]: Cause luxon do not interpolate `undefined` as input where moment(undefined) is simply moment(); (now)

File: addons/web/static/src/core/l10n/localization_service.js [hunk 6]
```diff
@@ -86,7 +86,7 @@ export const localizationService = {
             multiLang,
             thousandsSep: userLocalization.thousands_sep,
             weekStart: userLocalization.week_start,
-            code: language,
+            code: locale,
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/web/static/src/core/l10n/localization_service.js:89]: How ... Actually, I've added code in localization to use it later to replace legacy localization (there was "code" in database.parameters). 
@mcm-odoo use it here : https://github.com/odoo/odoo/pull/133153/files#diff-851c304e3f2a64138c1ff0f3de9856184dbfa0114da3fd76af0309eabe72043eR251-R263

File: addons/website_event_track/static/src/js/event_track_timer.js [hunk 7]
```diff
@@ -69,13 +69,13 @@ publicWidget.registry.websiteEventTrackTimer = publicWidget.Widget.extend({
...
      * the time remaining before the counter expires. The function will use
-     * MomentJS to transform the remaining time in more a human friendly format
+     * Luxon to transform the remaining time in more a human friendly format
      * Example: "in 32 minutes", "in 17 hours", etc.
      * @param {integer} remainingMs - Time remaining before the counter expires (in ms).
      */
     _updateTimerDisplay: function (remainingMs) {
         let $timerTextEl = this.$el.find('span');
-        let str = moment.duration(remainingMs, 'ms').humanize(true);
+        const str = "in " + formatDuration(remainingMs / 1000, true);
```
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/website_event_track/static/src/js/event_track_timer.js:78]: Damn -_-

File: addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pipu-odoo) [reply addons/point_of_sale/static/src/app/screens/ticket_screen/ticket_screen.js:378]: const value = order.validation_date ? DateTime.fromISO(order.validation_date) : DateTime.now();

File: addons/pos_restaurant/static/src/overrides/models/models.js [hunk 8]
```diff
@@ -29,7 +30,9 @@ patch(Order.prototype, {
         super.init_from_JSON(...arguments);
         if (this.pos.config.module_pos_restaurant) {
             this.tableId = json.table_id;
-            this.validation_date = moment.utc(json.creation_date).local().toDate();
+            const format = "yyyy-MM-dd HH:mm:ssZZ";
+            const creation_date = json.creation_date || DateTime.now().toFormat(format); //ex : json.creation_date = "2023-08-25 13:24:16+02:00"
+            this.validation_date = DateTime.fromFormat(creation_date, format).toJSDate();
```
Comment (brboi) [inline addons/pos_restaurant/static/src/overrides/models/models.js:35]:
```suggestion
            const creationDate = parseDateTime(json.creation_date) || DateTime.now();
            this.validation_date = creationDate.toJSDate();
```

This should work.

**Review by brboi — COMMENTED** [review]
LGTM ;-)

**Review by brboi — APPROVED** [review]
robodoo r+

**Pull request closed without merging**